discovery_engine_core/web-api/src/handlers.rs

```diff
@@ -149,35 +150,46 @@ pub(crate) async fn handle_user_interactions(
     body: InteractionRequestBody,
     state: Arc<AppState>,
 ) -> Result<impl warp::Reply, Rejection> {
-    let documents = state
+    let ids = body
+        .documents
+        .iter()
+        .map(|document| &document.document_id)
+        .collect::<Vec<_>>();
+    let embeddings = state
         .elastic
-        .get_documents_by_ids(&[body.document_id])
+        .get_documents_by_ids(&ids)
         .await
-        .map_err(handle_elastic_error)?;
+        .map_err(handle_elastic_error)?
+        .into_iter()
+        .map(|document| (document.id, document.embedding))
+        .collect::<HashMap<_, _>>();
 
-    if let Some(document) = documents.first() {
-        match body.user_interaction {
+    if embeddings.len() < ids.iter().unique().count() {
+        debug!("Document not found");
+        return Ok(StatusCode::NOT_FOUND);
+    }
+
+    for document in body.documents {
```

Comment:
🌱  Later, it may make sense to first update the cois locally for all interactions and then do a single update to reduce the number of transactions between the api and postgres